broken-access-control.1
.\" Automatically generated by Pandoc 2.17.1.1
.\"
.\" Define V font for inline verbatim, using C font in formats
.\" that render this, and otherwise B font.
.ie "\f[CB]x\f[]"x" \{\
. ftr V B
. ftr VI BI
. ftr VB B
. ftr VBI BI
.\}
.el \{\
. ftr V CR
. ftr VI CI
. ftr VB CB
. ftr VBI CBI
.\}
.TH "BROKEN-ACCESS-CONTROL" "1" "20 Jun 2023" "v0.1" "Tool for testing access control - documentation"
.hy
.SH NAME
.PP
broken-access-control - script to test access control on a web app
during a pentest
.SH SYNOPSIS
.TP
broken-access-control [-h] [-c CONFIG] [\[en]login-path LOGIN_PATH] [\[en]id-field ID_FIELD]
.nf
\f[C]
               [--pwd-field PWD_FIELD] [--login-code LOGIN_CODE | --login-text LOGIN_TEXT] [--csrf]
\f[R]
.fi
.nf
\f[C]
               [--csrf-class CSRF_CLASS] [--csrf-name CSRF_NAME] [-o OUT] [-w WAIT]
\f[R]
.fi
.nf
\f[C]
               [--limit-users LIMIT_USERS] [--limit-urls LIMIT_URLS] [-m MAX_RETRIES] [-t TIMEOUT] [-p] [-v] [-d] [-j]
\f[R]
.fi
.nf
\f[C]
               [--allow-redirect] [--disable-unauth] [--disable-https]
\f[R]
.fi
.nf
\f[C]
               [host] [users] [urls]
\f[R]
.fi
.SH DESCRIPTION
.PP
With a given list of users and URLs and a host server, test all accesses
and store result as a table or in JSON format.
For the moment, this tool return only the HTTP status code of the
response.
Arguments can be specified from the command line or from a configuration
file (see: FILES).
.PP
The script will first send a GET request to all the URLs in the given
file without any authentication (unless \f[B]disable-unauth\f[R] is
set).
Then it will try to login to the application with the given parameters
(\f[B]login-path\f[R], \f[B]id-field\f[R], \f[B]pwd-field\f[R],
\f[B]login-code\f[R], \f[B]login-text\f[R]).
To authenticate the script send a POST request to the page given in
\f[B]login-path\f[R].
The payload is created with the \f[B]id-field\f[R] and
\f[B]pwd-field\f[R] parameters and with the values parsed from the CSV
file containing the users credentials (\f[B]users\f[R] parameters).
.PP
It\[cq]s possible to use a CSRF mechanism to connect as a classic user
do.
You must enable CSRF using with \f[B]csrf\f[R] parameter and then, you
can give the type of HTML tag which contains the CSRF token with
\f[B]csrf-class\f[R] and their name with \f[B]csrf-name\f[R].
.PP
\f[B]Example:\f[R]
.PP
With default parameters (see: OPTIONS), the following request will be
send for logging in:
.IP
.nf
\f[C]
POST /login HTTP/1.1
[... Headers ...]

email=user1&password=password1
\f[R]
.fi
.PP
And then, by default the script detect a valid login attempt with a HTTP
200 response, but you can change the status code, or detect a valid
authentication with a string in the response text.
By default, the tool does not follow redirections, for example, to
detect a valid authentication upon a 302 status code in HTTP response,
but you can enable the redirection.
.SH OPTIONS
.TP
-h, \[en]help
show help message and exit
.SS Load arguments from configuration file:
.TP
-c CONFIG, \[en]config CONFIG
configuration file to load arguments from, JSON format, override command
line arguments
.SS Required arguments without configuration file:
.TP
host
specify the target host
.TP
users
file containing the list of users
.TP
urls
file containing the list of URLs
.SS Login helper arguments:
.TP
\[en]login-path LOGIN_PATH
path to the login page of the application (default: /login)
.TP
\[en]id-field ID_FIELD
name of the id field in the request for login (default: email)
.TP
\[en]pwd-field PWD_FIELD
name of the password field in the request for login (default: password)
.TP
\[en]login-code LOGIN_CODE
status code upon a valid login request (default: 200)
.TP
\[en]login-text LOGIN_TEXT
text in the page upon a valid upon login request
.TP
\[en]csrf
enable use of a CSRF token for login (default: False)
.TP
\[en]csrf-class CSRF_CLASS
class of the CSRF token element in the login page (default: input)
.TP
\[en]csrf-name CSRF_NAME
name of the CSRF token element in the login page (default: csrf)
.SS Optional arguments:
.TP
-o OUT, \[en]out OUT
store the output in this file
.TP
-w WAIT, \[en]wait WAIT
waiting time to limit requests rate, in seconds (default: 1)
.TP
\[en]limit-users LIMIT_USERS
set the limit of users to test
.TP
\[en]limit-urls LIMIT_URLS
set the limit of URLs to test
.TP
-t TIMEOUT, \[en]timeout TIMEOUT
timeout for each HTTP request (default: 5)
.TP
-m MAX_RETRIES, \[en]max-retries MAX_RETRIES
max retries attempts if request failed (default: 3)
.TP
-p, \[en]proxy
enable use of proxy, for the moment: 127.0.0.1:8080 (default: False)
.TP
-v, \[en]verbose
see info and warning messages (default: False)
.TP
-d, \[en]debug
see debug messages (default: False)
.TP
-j, \[en]json
save results in JSON format (default: False)
.TP
\[en]allow-redirect
follow requests redirection (default: False)
.TP
\[en]disable-unauth
disable unauthenticated tests (default: False)
.TP
\[en]disable-https
disable HTTPS (default: False)
.SH FILES
.SS Users file
.PP
The file containing the users credentials must fit the following format,
including \[lq]\f[V];\f[R]\[rq] delimiter:
.IP
.nf
\f[C]
role;username;password
user;johndoe;Password#123
administrator;admin;P\[at]ssw0rd456
\f[R]
.fi
.SS Configuration file
.PP
Parameters can be load from a configuration file.
Here is an example of this file :
.IP
.nf
\f[C]
{
  \[dq]host\[dq]: \[dq]example.domain.fr\[dq],
  \[dq]users\[dq]: \[dq]user.csv\[dq],
  \[dq]urls\[dq]: \[dq]url.txt\[dq],
  \[dq]login_path\[dq]: \[dq]/login\[dq],
  \[dq]id_field\[dq]: \[dq]email\[dq],
  \[dq]pwd_field\[dq]: \[dq]password\[dq],
  \[dq]login_code\[dq]: 200,
  \[dq]login_text\[dq]: null,
  \[dq]csrf\[dq]: false,
  \[dq]csrf_class\[dq]: \[dq]input\[dq],
  \[dq]csrf_name\[dq]: \[dq]csrf\[dq],
  \[dq]out\[dq]: null,
  \[dq]wait\[dq]: 1,
  \[dq]limit_users\[dq]: null,
  \[dq]limit_urls\[dq]: null,
  \[dq]timeout\[dq]: 5,
  \[dq]max_retries\[dq]: 3,
  \[dq]proxy\[dq]: false,
  \[dq]verbose\[dq]: false,
  \[dq]debug\[dq]: false,
  \[dq]json\[dq]: false,
  \[dq]allow_redirect\[dq]: false,
  \[dq]disable_unauth\[dq]: false,
  \[dq]disable_https\[dq]: false
}
\f[R]
.fi
.SH TO DO
.IP \[bu] 2
Add possibility of test control access with POST requests with
parameters
.IP \[bu] 2
Possibility of adding headers to the requests
.IP \[bu] 2
Auto detection of login method (in progress)
.SH AUTHORS
Louka.


broken-access-control.py
#!/usr/bin/python3

from urllib.parse import urlparse
from pwn import log, context
from json import dump, load, dumps
from csv import DictReader
from time import sleep

import sys
import argparse
import usersession as us

class BrokenAccessControl():
    # Constants
    UNAUTHENTICATED = "Unauth. user"
    # Users CSV fields name
    descField = "role"
    userField = "username"
    pwdField  = "password"

    # Constructor
    def __init__(self, argsDict):
        self.__dict__.update(argsDict)
        DEFAULT_PROXY = {
            "http":  "127.0.0.1:8080",
            "https": "127.0.0.1:8080"
        }
        self.proxy = DEFAULT_PROXY if self.proxy else None

    # open and extract content of URLs file
    def loadURLs(self):
        urlsFile = self.urls
        with open(urlsFile, 'r') as urlf:
            # read url list, split by line and remove last element (empty)
            content = urlf.read().split('\n')[:-1]
        urls = list()
        for url in content:
            p = urlparse(url)
            urls.append(p.path + ('?' + p.query if p.query != '' else ''))
        return urls

    # open and extract content of CSV users file
    def loadUsers(self):
        usersFile = self.users
        users = list()
        with open(usersFile, 'r') as userf:
            userReader = DictReader(userf, delimiter=';')
            # load users (list of dictionnaries)
            for user in userReader:
                users.append(user)
        return users

    # Test all access controls with a GET request to each urls for each user
    def testAll(self, users, urls):
        # init result
        result = dict()
        for url in urls:
            result[url] = dict()
        # Add unauthenticated user if needed
        if not self.disable_unauth:
            users = [{
                self.descField: self.UNAUTHENTICATED,
                self.userField: None,
                self.pwdField: None
            }] + users
        # Send all requests for each users
        for user in users:
            log.info("Testing %d URLs for user: %s, role: %s" % (len(urls), user[self.userField], user[self.descField]))
            session = us.UserSession(self.host,
                self.login_path, self.id_field, self.pwd_field,
                self.login_code, (self.login_text is not None), self.login_text,
                self.csrf, self.csrf_name, self.csrf_class,
                self.disable_https, self.allow_redirect, self.timeout, self.proxy)
            # Try to login
            if user[self.descField] is not self.UNAUTHENTICATED:
                session.login(user[self.userField], user[self.pwdField])
                # Test if connection suceed
                if not session.isConnected:
                    log.warn("Skipping user %s: can't log in" % user[self.userField])
                    continue
            # Add sleep time to reduce request rate
            sleep(self.wait)
            for url in urls:
                try:
                    r = session.get(url)
                    count = 0
                    while r.status_code == 502 and count < self.max_retries:
                        log.debug(f"Got 502 when trying to access {url}, retrying")
                        r = session.get(url)
                        count += 1
                        sleep(self.wait)
                    result[url][user[self.descField]] = r.status_code
                # request timeout
                except requests.exceptions.Timeout:
                    log.debug(f"Request to {url} timeout")
                    result[url][user[self.descField]] = "Timeout"
            sleep(self.wait * 3)
        return result

    # Save result as text or in JSON format
    def save(self, result):
        output = sys.stdout
        # open output file
        if self.out is not None:
            try:
                output = open(self.out, 'w')
            except:
                log.warn("Can't open given output file, write to stdout")
        # Save in JSON format
        if self.json:
            dump(result, output, indent=4)
            if output == sys.stdout:
                print()
        # Save in human-readable format
        else:
            self.writeTable(result, output)
        # Close output file if opened
        if self.out is not None:
            output.close()

    # Write data as a formatted table in out
    def writeTable(self, data, out):
        # determine padding
        maxPad = max([len(path) for path in data.keys()])
        users = list(data.values())[0].keys()
        columnLength = max([len(u) for u in users]) + 1
        # print users
        fmtStr = ("{:<%d} " % maxPad) + ("{:<%d}" % columnLength) * len(users) + '\n'
        out.write(fmtStr.format("URLs", *users))
        out.write('-' * (maxPad + 1 + columnLength * len(users)) + '\n')
        # print table
        for path, codes in zip(data.keys(), data.values()):
            fmtStr = ("{:<%d} " % maxPad) + ("{:<%d}" % columnLength) * len(codes) + '\n'
            out.write(fmtStr.format(path, *codes.values()))
        
    # Run the broken access control test
    def run(self):
        # Host informations
        log.info(f"Host: {self.host}")
        # Proxy informations
        if self.proxy is not None:
            log.info("HTTP proxy is  : %s" % self.proxy['http'])
            log.info("HTTPS proxy is : %s" % self.proxy['https'])
        try:
            # Load users
            log.info(f"Load users file: {self.users}")
            users = self.loadUsers()
            # Load URLs
            log.info(f"Load target URLs file: {self.urls}")
            log.warn("WARNING: ensure /logout or equivalent is not in URLs or at the end of the list")
            urls  = self.loadURLs()
        except:
            log.error("Can't load or open users or URLs file")
        # Limits the numbers of users and URLs to test if asked (debug)
        if self.limit_users is not None:
            users = users[:self.limit_users]
        if self.limit_urls is not None:
            urls = urls[:self.limit_urls]
        # Start test
        result = self.testAll(users, urls)
        self.save(result)
        # Finished
        log.info("Done")


# Entry point
if __name__ == "__main__":
    # Parsing command line
    parser = argparse.ArgumentParser(
        prog='broken-access-control',
        description="""
            With a given list of users and URLs and a host server, test all accesses and store result as a table or in JSON format.
            Arguments can be specified from the command line or from a configuration file (see: template-config.json)
            """,
        formatter_class=argparse.ArgumentDefaultsHelpFormatter
    )
    # Load arguments from configuration file
    cfg = parser.add_argument_group('Load arguments from configuration file')
    cfg.add_argument('-c', '--config', type=str, help="configuration file to load arguments from, JSON format, override command line arguments")
    # Required arguments
    req = parser.add_argument_group('Required arguments without configuration file')
    req.add_argument("host",  type=str, help="specify the target host", nargs='?')
    req.add_argument("users", type=str, help="file containing the list of users", nargs='?')
    req.add_argument("urls",  type=str, help="file containing the list of URLs", nargs='?')
    # Login helper arguments
    login = parser.add_argument_group('Login helper arguments')
    login.add_argument("--login-path",  type=str, help="path to the login page of the application", default="/login")
    login.add_argument("--id-field",    type=str, help="name of the id field in the request for login", default="email")
    login.add_argument("--pwd-field",   type=str, help="name of the password field in the request for login", default="password")
    loginValidation = login.add_mutually_exclusive_group()
    loginValidation.add_argument("--login-code", type=int, help="status code upon a valid login request", default=200)
    loginValidation.add_argument("--login-text", type=str, help="text in the page upon a valid upon login request")
    login.add_argument("--csrf", action="store_true", help="name of the field in the POST request for login")
    login.add_argument("--csrf-class", type=str, help="class of the CSRF token element in the login page", default="input")
    login.add_argument("--csrf-name",  type=str, help="name of the CSRF token element in the login page", default="csrf")
    # Optional arguments
    options = parser.add_argument_group('Optional arguments')
    options.add_argument("-o", "--out",   type=str, help="store the output in this file")
    options.add_argument("-w", "--wait",  type=int, help="waiting time to limit requests rate, in seconds", default=1)
    options.add_argument("--limit-users", type=int, help="set the limit of users to test")
    options.add_argument("--limit-urls",  type=int, help="set the limit of URLs to test")
    options.add_argument("-t", "--timeout", type=int, help="timeout for each HTTP request", default=5)
    options.add_argument("-m", "--max-retries", type=int, help="max retries attempts if request failed", default=3)
    options.add_argument("-p", "--proxy",    action="store_true", help="enable use of proxy, for the moment: 127.0.0.1:8080")
    options.add_argument("-v", "--verbose",  action="store_true", help="see info and warning messages")
    options.add_argument("-d", "--debug",    action="store_true", help="see debug messages")
    options.add_argument("-j", "--json",     action="store_true", help="save results in JSON format")
    options.add_argument("--allow-redirect", action="store_true", help="follow requests redirection")
    options.add_argument("--disable-unauth", action="store_true", help="disable unauthenticated tests")
    options.add_argument("--disable-https",  action="store_true", help="disable HTTPS")
    # If no arguments are provided, show help
    if len(sys.argv) == 1:
        parser.print_help()
        sys.exit(1)
    # Parse arguments
    args = parser.parse_args()
    argsDict = vars(args)
    if args.config is not None:
        try:
            configArgs = load(open(args.config))
            argsDict.update(configArgs)
        except:
            log.error("Configuration file loading failed")
    # Set verbosity level
    context.log_level = "debug" if argsDict["debug"] else ("info" if argsDict["verbose"] else "error")
    # Show args dict if debugging is enable
    log.debug(f"Arguments: {dumps(argsDict)}")
    # Check required arguments
    if args.config or (args.host and args.users and args.urls):
        # Init and run the test
        BrokenAccessControl(argsDict).run()
    else:
        log.error("Arguments host, users, and urls are required if no configuration file are provided")
        sys.exit(1)
